### ISSUE TITLE
3.0 + TP2 - issue with sequences when multiple vertices are created with graph API

### ISSUE THREAD

**Issue by laa** [orientechnologies/orientdb#7421]
### OrientDB Version: 3.0

The bug is described there http://stackoverflow.com/questions/43993477/orientdb-3-0-issue-with-sequences-when-multiple-vertices-are-created-with-graph . 

One issue is fixed could you also update related SO entry?

**Comment by careerscale** [comment]
I haven't used embeeded db (answer the question in stackoverflow). But learn and do it.

**Comment by robfrank** [comment]
I forked your project to do some tests.

**Comment by careerscale** [comment]
Ok, pls let me know  if anything else needed from me.

**Comment by robfrank** [comment]
I tested against the latest snapshot, using your tests, and all works fine. As a double check, can you move on 3.0.0-SNAPSHOT on maven pom AND server? 
You can donwload the latest community tar.gz from https://oss.sonatype.org/content/repositories/snapshots/com/orientechnologies/orientdb-community/3.0.0-SNAPSHOT/

Moreover, I moved the code to orientdb-gremlin (thinkerPop 3.0) and it works fine.

**Comment by robfrank** [comment]
I tested against the latest snapshot, using your tests, and all works fine. As a double check, can you move on 3.0.0-SNAPSHOT on maven pom AND server? 
You can donwload the latest community tar.gz from https://oss.sonatype.org/content/repositories/snapshots/com/orientechnologies/orientdb-community/3.0.0-SNAPSHOT/

Moreover, I moved the code to orientdb-gremlin (thinkerPop 3.0) and it works fine. 

To enable snapshots, just add this to your pom:
```xml
  <properties>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
        <orientdb-version>3.0.0-SNAPSHOT</orientdb-version>
    </properties>
    <repositories>
        <repository>
            <id>sonatype-nexus-snapshots</id>
            <name>Sonatype Nexus Snapshots</name>
            <url>https://oss.sonatype.org/content/repositories/snapshots</url>
            <releases>
                <enabled>false</enabled>
            </releases>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
    </repositories>

```

**Comment by careerscale** [comment]
I think I am using the latest server (building locally for "develop" branch

but from application perspective, yes, I will use this and see if it works fine for me too.

Thanks

**Comment by careerscale** [comment]
Not really, I did see issues still. the uniqueness is not maintained. built latest server code.

Added assertion for uniqueness, could you please check again?

**Comment by robfrank** [comment]
I just made a PR to your project. I migrated the test to tinker pop 3 API, and it works fine. clean the database before running the test with: 

```
DELETE VERTEX User
```

I also suggest to create a unique index on User's id: it will speed up lookup on the id property.
```
CREATE INDEX User.id on User(id) UNIQUE
```

Take a look to
- http://orientdb.com/docs/3.0.x/sql/SQL-Create-Index.html
- http://orientdb.com/docs/3.0.x/indexing/Indexes.html

**Comment by careerscale** [comment]
I ran again, and the tests fail nicely. There are dupicate id values generated, there are exceptions thrown resulting in failures too.

Are you doing any change to the code or pom.xml? I have updated the dependencies as you suggested (3.0 snapshot) yesterday but I have no luck

**Comment by careerscale** [comment]
![graph api id generation issue](https://cloud.githubusercontent.com/assets/342691/26144655/877d8784-3a9e-11e7-8807-adf8da35d429.PNG)

**Comment by careerscale** [comment]
Sorry, yah, your test class works fine. what was the different thing you made?

**Comment by careerscale** [comment]
Should we use gremlin API? is that TP3?  is the graph api is buggy?

**Comment by lvca** [comment]
@careerscale If you're using OrientDB 3.0 with an app developed from scratch, I won't use TP2.6 but just the MultiModel API. Do you have this issue with plain Multi-Model API?

**Comment by careerscale** [comment]
yes, this bug is about the issue with multi model api (we are using graphdb api) dependency is given below.

			<dependency>
				<groupId>com.orientechnologies</groupId>
				<artifactId>orientdb-graphdb</artifactId>
				<version>${orientdb-version}</version>
			</dependency>
with 3.0 snapshots

**Comment by robfrank** [comment]
To be more clear, on 3.0 we provide 3 sets of different api
- multi-model api : http://orientdb.com/docs/3.0.x/java/Java-MultiModel-API.html, you need only orientdb-core
- tinkerPop/gremlin 3.0 api: http://orientdb.com/docs/3.0.x/tinkerpop3/OrientDB-TinkerPop3.html
```xml
		<dependency>
			<groupId>com.orientechnologies</groupId>
			<artifactId>orientdb-gremlin</artifactId>
			<version>${orientdb-version}</version>
		</dependency>

```
- tikerPop/gremlin 2.6 (for backward compatibility):  http://orientdb.com/docs/3.0.x/java/Graph-Database-Tinkerpop.html:
```xml
		<dependency>
			<groupId>com.orientechnologies</groupId>
			<artifactId>orientdb-graphdb</artifactId>
			<version>${orientdb-version}</version>
		</dependency>

```

The multi-model API allows to work with graph and document in a unified, flexible api, without additional external dependencies.

As stated by @lvca , our suggestion is to rely on Multimodel if you don't need gremlin, or at least on gremlin 3.0 (orientdb-gremlin module) if you need gremlin support.   

I'm going to provide a rewritten version of the test that uses the multimodel-api just to show how it works, so you can choose.

**Comment by careerscale** [comment]
Ok, there is a realization now, I was thinking the order of APIs was multi model bering the core and graph api being the 3.0 special feature while gremlin one being the 2.6 TP, but  I was wrong.

Quick question:

With graph API, I got this API, 
OrientGraph.getVertices(Plan.PLAN, keys, values);

what is the equalant one with gremlin? I think we are more confident using the pure graph model than the document model.

It turns out to be that graphdb api is bit of a legacy one resulting in some of my observations, so I would like to move to gremlin one asap.

Thanks for this information, it is great to learn the differences

**Comment by careerscale** [comment]
Ok, I think I got it,, thanks. your sample unit tests was good enough for me.
OResultSet vertices = graph.executeSql("Select from User", params);
        List<Long> ids = Lists.newArrayList();
        vertices.vertexStream().forEach(v -> {
        }
        );

**Comment by careerscale** [comment]
Should we mark this ticket  as resolved?  we are moving to gremlin 3.0 snapshot.

**Comment by robfrank** [comment]
ok, closing. thank you very much, the sandbox project is a very good idea!
